transport/grpc/outbound_test.go

```diff
@@ -200,3 +203,159 @@ func TestCallStreamWithInvalidPeer(t *testing.T) {
 		}.Error(),
 	)
 }
+
+func TestCallSvcMatchSuccess(t *testing.T) {
+	matchServer := grpc.NewServer(
+		grpc.CustomCodec(customCodec{}),
+		grpc.UnknownServiceHandler(
+			func(srv interface{}, stream grpc.ServerStream) error {
+				ctx := stream.Context()
+				// Get transport request
+				md, ok := metadata.FromIncomingContext(ctx)
+				if md == nil || !ok {
+					return yarpcerrors.Newf(yarpcerrors.CodeInternal, "cannot get metadata from ctx: %v", ctx)
+				}
+				transportRequest, err := metadataToTransportRequest(md)
+				if err != nil {
+					return err
+				}
+
+				responseWriter := newResponseWriter()
+				defer responseWriter.Close()
+
+				// Echo accepted rpc-service in response header
+				responseWriter.AddSystemHeader(ServiceHeader, transportRequest.Service)
+
+				// Send the response attributes back and end the stream.
+				if sendErr := stream.SendMsg(responseWriter.Bytes()); sendErr != nil {
+					// We couldn't send the response.
+					return sendErr
+				}
+				if responseWriter.md != nil {
+					stream.SetTrailer(responseWriter.md)
+				}
+				return err
+			}),
+	)
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:15213"))
+	require.NoError(t, err)
+	go func() {
+		// TODO: do we have consistency issue here? Request comes before server begin to serve
```

Comment:
To solve this issue, we can use a channel. Let's 
1. create a channel before the goroutine
2. close it in the goroutine
3. wait for the channel to close after the go routine

This will ensure that the test only starts _after_ the server is listening.

Reply:
Thanks a lot for the detailed instructions. I just looked at the source code of gRPC, it seems like we don't need to handle the consistency issue. Because gRPC will grab a global lock before it starts the server, and unlock after it started, operations like add a connection will wait for this global lock when server not started.